internal/semver/semver.go

```diff
@@ -0,0 +1,64 @@
+package semver
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/go-version"
+	"github.com/hashicorp/vault/api"
+
+	"github.com/hashicorp/terraform-provider-vault/internal/consts"
+)
+
+func getTargetVaultVersion(ctx context.Context, client *api.Client) (string, error) {
+	resp, err := client.Sys().SealStatusWithContext(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	if resp == nil {
+		return "", fmt.Errorf("expected response data, got nil response")
+	}
+
+	if resp.Version == "" {
+		return "", fmt.Errorf("key %q not found in response", consts.FieldVersion)
+	}
+
+	return resp.Version, nil
+}
+
+// GreaterThanOrEqual receives a context, a Vault API client
+// and a minimum version that the Vault server version
+// should be above.
+//
+// It uses the go-version package
+// to perform a semantic version comparison, and
+// returns:
+//    - a boolean describing whether the Vault
+//      server version was above the minimum version
+//    - the current Vault server version as a string
+//    - errors captured during operation, if any
+//
+// This function can be used to perform semantic version comparisons
+// to conditionally enable features, or to resolve any diffs in the TF
+// state based on the Vault version.
+func GreaterThanOrEqual(ctx context.Context, client *api.Client, minVersionString string) (bool, string, error) {
```

Comment:
I wonder if we should be doing the initial version check in `provider.NewProviderMeta` instead. This would allow the provider to abort early if it encounters an error getting the Vault version. It would also make it easier to use in the resource code since we would no longer have to handle the error. We would store the `version.Version` in the `ProviderMeta` so there would be no need to return the vault version string.

If we make `minVersionString` a `version.Version` the caller would be responsible for handling any parse errors.

Reply:
@benashz I'm glad you brought this up! I initially thought of this approach as well, and the main drawback that came to my mind (which I remember we briefly chatted about) was in the case where users use the same TF instance to configure multiple versions of Vault. In that case, the approach of storing the current Vault version to the `ProviderMeta` would explicitly disallow the use of the TFVP in those kind of scenarios, and I wanted to get your take on the larger implications/architecture before going down that route. I do agree that the current implementation adds an overhead of multiple calls to the same endpoint

Reply:
I think we are okay here since each instance of the provider corresponds to a single Vault server instance.

Reply:
By instance of the provider, I am referring to each `provider{}` block defined; terraform will spawn a separate process for each block defined.